drivers/timer/arcv2_timer0.c

```diff
@@ -211,11 +270,18 @@ int z_clock_driver_init(struct device *device)
 	start_time = last_time;
 #else
 	last_load = CYC_PER_TICK;
+	overflow_cyc = 0;
+	announced_cycles = 0;
 
 	IRQ_CONNECT(IRQ_TIMER0, CONFIG_ARCV2_TIMER_IRQ_PRIORITY,
 		    timer_int_handler, NULL, 0);
 
 	timer0_limit_register_set(last_load - 1);
+	/* make timer irq pulse sensitive, and self-clear when it's handled
+	 * then we can clear the IP bit of CTRT in non-interrupt context,
+	 * without missing timer irq.
+	 */
+	z_arc_v2_irq_unit_sensitivity_set(IRQ_TIMER0, 1);
```

Comment:
If you specify an IRQ line mode instead of meaningless `1` it's better to have `ARCV2_IRQ_MODE_PULSE` etc. Otherwise this very special setup (I never saw configuration of ARC Timer done via Pulse mode before so won't be surprised if it might produce unexpected results) might not attract attention of the next reader.

Also I don't quite understand the whole idea of switching to pulse-triggered IRQ from the ARC Timer. Which problem does it solve? And if it is still usable with all our current HW and simulation platforms. I'd strongly suggest to not use pulse-triggered IRQ source if possible.

Reply:
I run the tests on nsim_em/nsim_sem/nsim_hs,  the results shows pulse trigger is ok. If real hardware is also ok, i think there will be no problem.

Why we need pulse trigger?
Because we need to update the overflow_cyc  in elapsed() when irq is locked. Without pulse trigger, the clear of IP bit will also clear the irq request, then when irq is unlocked, the timer irq will be lost, then no announcement of systick.  I think this is unacceptable for RTOS.

I have also considered to use software irq to keep the timer irq, but it will make things more complex.

The ARCv2 ISA also allows pulse trigger to do a self-clear of IRQ.

For wall clock , it relies on extra hardware. We can use one 32bit timer to do the same thing as ARM SysTick (24bit) does.

Reply:
So I went once again over long discussion in https://github.com/zephyrproject-rtos/zephyr/pull/24332 and now is confused even more.

Yet I don't see any justification for that dark (or rather dirty) magic with IRQs except we may `busy_wait()` twice longer. But do we really want to trade that IMHO meaningless twice longer `busy_wait()` for:
1. More tricky driver logic
2. "Delayed" IRQ handling

I'd propose to just add a check in `busy_wait()` for too long periods. Moreover I think long `busy_wait()` which implicitly means spinning w/ all [not only timer] IRQs disabled is something you don't want to happen in good RTOS-based solution as responsiveness of it will be quite slow in some cases :)

So first I'd like to figure out if we're trying to solve a valid problem here or we're overengineering something which has no real use in a sane software design.

Maybe we even return to https://github.com/zephyrproject-rtos/zephyr/pull/24332 and continue discussion there before drafting fixes. @andyross @andrewboie what do you think?